File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        int[] task = new int[n + 1];
        int[] profit = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            task[i] = Integer.parseInt(st.nextToken());
            profit[i] = Integer.parseInt(st.nextToken());
        }

        int[] dp = new int[n + 2];

        for (int day = n; day > 0; day--) {
            int finishDay = day + task[day];

            if (finishDay > n + 1) {
                dp[day] = dp[day + 1];
            } else {
                dp[day] = Math.max(dp[day + 1], profit[day] + dp[finishDay]);
            }
        }

        System.out.println(dp[1]);
    }
}

File: 백준/Silver/14501. 퇴사/퇴사.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val task = IntArray(n + 1)
    val profit = IntArray(n + 1)

    repeat(n) { i ->
        val (t, p) = br.readLine().split(" ").map { it.toInt() }
        task[i + 1] = t
        profit[i + 1] = p
    }

    val dp = IntArray(n + 2)

    for (day in n downTo 1) {
        val finishDay = day + task[day]
        dp[day] = if (finishDay > n + 1) {
            dp[day + 1]
        } else {
            maxOf(dp[day + 1], profit[day] + dp[finishDay])
        }
    }

    println(dp[1])
}
